src/QsCompiler/CompilerOptimizations/StatementRemoving.fs

```diff
@@ -36,12 +40,35 @@ type internal StatementRemover() =
             jointFlatten (s.Lhs, s.Rhs) |> Seq.map (QsValueUpdate.New >> QsValueUpdate)
         | QsQubitScope s when isAllDiscarded s.Binding.Lhs ->
             s.Body.Statements |> Seq.map (fun x -> x.Statement)
-        | QsScopeStatement s -> s.Body.Statements |> Seq.map (fun x -> x.Statement)
-        | _ when not c.hasQuantum && not c.hasMutation && not c.hasInterrupts -> Seq.empty
+        | QsQubitScope s ->
+            let mutable newStatements = []
+            let myList = jointFlatten (s.Binding.Lhs, s.Binding.Rhs) |> Seq.collect (fun (l, r) ->
+                match l, r.Resolution with
+                | VariableName name, QubitRegisterAllocation {Expression = IntLiteral num} ->
+                    let elemI = fun i -> Identifier (LocalVariable (NonNullable<_>.New (sprintf "__qsItem%d__%s__" i name.Value)), Null)
+                    let expr = Seq.init (safeCastInt64 num) (elemI >> wrapExpr Qubit) |> ImmutableArray.CreateRange |> ValueArray |> wrapExpr (ArrayType (ResolvedType.New Qubit))
+                    let newStmt = QsVariableDeclaration (QsBinding.New QsBindingKind.ImmutableBinding (l, expr))
+                    newStatements <- wrapStmt newStmt :: newStatements
+                    Seq.init (safeCastInt64 num) (fun i ->
+                        VariableName (NonNullable<_>.New (sprintf "__qsItem%d__%s__" i name.Value)),
+                        ResolvedInitializer.New SingleQubitAllocation)
+                | DiscardedItem, _ -> Seq.empty
+                | _ -> Seq.singleton (l, r)) |> List.ofSeq
+            match myList with
+            | [] -> newScopeStatement s.Body |> Seq.singleton
+            | [lhs, rhs] ->
+                let newBody = QsScope.New (s.Body.Statements.InsertRange (0, newStatements), s.Body.KnownSymbols)
+                QsQubitScope.New s.Kind ((lhs, rhs), newBody) |> QsQubitScope |> Seq.singleton
+            | _ ->
+                let lhs = List.map fst myList |> ImmutableArray.CreateRange |> VariableNameTuple
+                let rhs = List.map snd myList |> ImmutableArray.CreateRange |> QubitTupleAllocation |> ResolvedInitializer.New
+                let newBody = QsScope.New (s.Body.Statements.InsertRange (0, newStatements), s.Body.KnownSymbols)
```

Comment:
Are the known symbols listed still accurate? I believe you may need to adapt that, same above.

Reply:
There are a lot of places in the code that can change the symbols visible at the top of a scope. For instance, inlining a callable will add all the symbols of the parent function to that callable, or removing an unused variable will remove symbols from all following scopes. Instead of trying to update this metadata in each transformation, which seems difficult and prone to error, could we entirely regenerate it after running all the optimization steps?

Reply:
Yes, I think that would be fine. I suggest to explicitly strip that information everywhere entirely then. Could you make sure that info is stripped in general? And in particular it needs to be documented. I would say stripping that info cleanly and documenting the behavior is the important part, even if you don't get around to writing a transformation that reconstructs that info.